libraries/ms-common/src/x86/adc.c

```diff
@@ -1,23 +1,172 @@
+#include <stddef.h>
+
 #include "adc.h"
+#include "interrupt.h"
+#include "soft_timer.h"
+// x86 implementation very similar to STM32F0 implementation.
+// adc_read_raw should always return 4090.
+// Vdda locked at 3300 mV.
+// adc_read_converted should always return close to 2V
+
+#define ADC_RETURNED_VOLTAGE_RAW 2500
+#define ADC_CONTINUOUS_CB_FREQ_MS 50
+
+// TS_CAL addresses obtained from section 3.10.1 of the specific device
+// datasheet
+#define ADC_TS_CAL1 0x1FFFF7b8
+#define ADC_TS_CAL2 0x1FFFF7c2
+
+// ADC_VREFINT_CAL address obtained from section 3.10.2 of the specific device
+// datasheet
+#define ADC_VREFINT_CAL 0x1FFFF7ba
+
+typedef struct AdcInterrupt {
+  AdcCallback callback;
+  void *context;
+  uint16_t reading;
+} AdcInterrupt;
+
+static AdcInterrupt s_adc_interrupts[NUM_ADC_CHANNELS];
+
+static bool s_active_channels[NUM_ADC_CHANNELS];
+
+// Formula obtained from section 13.9 of the reference manual. Returns reading
+// in kelvin
+static uint16_t prv_get_temp(uint16_t reading) {
+  uint16_t ts_cal1 = *(uint16_t *)ADC_TS_CAL1;
+  uint16_t ts_cal2 = *(uint16_t *)ADC_TS_CAL2;
```

Comment:
You seem to be dereferencing a hardcoded pointer on x86? This will fail with a segfault, the `ts_cal` values themselves should probably be hardcoded like Vdda. I'm surprised there are no tests for temp/vdda.